core/message_pool.go

```diff
@@ -149,7 +178,7 @@ func collectChainsMessagesToHeight(ctx context.Context, store *hamt.CborIpldStor
 // TODO there is considerable functionality missing here: don't add
 //      messages that have expired, respect nonce, do this efficiently,
 //      etc.
-func UpdateMessagePool(ctx context.Context, pool *MessagePool, store *hamt.CborIpldStore, old, new types.TipSet) error {
+func (pool *MessagePool) UpdateMessagePool(ctx context.Context, store *hamt.CborIpldStore, old, new types.TipSet) error {
```

Comment:
Blocking: I do not think this functionality should be tightly tied to the message pool. I have it as a low-pri cleanup to move the integration with the chain store out of this package, so that message_pool is a simple data structure, and these algorithms use its public interface. Please revert the receiver change so that the move is easier later.

See my top-level comment about moving the timestamp to a parameter of `Add`.

Reply:
See comment about the role of `MessagePool`